datadog/reporter.go

```diff
@@ -0,0 +1,296 @@
+)
+
+const (
+	// DefaultBufferSize contains the number of metrics the datadog reporter
+	// will capture before forcing a flush
+	DefaultBufferSize = 512
+)
+
+const (
+	typeGauge   = "gauge"
+	typeRate    = "rate"
+	typeCounter = "counter"
+	typeTimer   = "timer"
+)
+
+var (
+	poolMetric = &sync.Pool{
+		New: func() interface{} {
+			return &metric{}
+		},
+	}
+)
+
+type metric struct {
+	Name   string      `json:"metric"`
+	Points [][]float64 `json:"points"`
+	Type   string      `json:"type"`
+	Host   string      `json:"host,omitempty"`
+	Tags   []string    `json:"tags,omitempty"`
+}
+
+func (m *metric) Reset() {
+	m.Name = ""
+	m.Points = nil
+	m.Type = ""
+	m.Host = ""
+	m.Tags = nil
+}
+
+type series struct {
+	Metrics []*metric `json:"series"`
+}
+
+type Reporter struct {
+	metrics     []*metric
+	bufSize     int
+	offset      int
+	handlerFunc func(*http.Request) (*http.Response, error)
+	output      io.Writer
+	mux         *sync.Mutex
+	endpoint    string
+}
+
+// Reporting returns whether the reporter has the ability to actively report.
+func (r *Reporter) Reporting() bool {
+	return false
+}
+
+// Tagging returns whether the reporter has the capability for tagged metrics.
+func (r *Reporter) Tagging() bool {
+	return true
+}
+
+// Capabilities returns the capabilities description of the reporter.
+func (r *Reporter) Capabilities() tally.Capabilities {
+	return r
+}
+
+// post encoded metrics to datadog
+func (r *Reporter) post(data []byte) error {
+	req, err := http.NewRequest("POST", r.endpoint, bytes.NewReader(data))
+	if err != nil {
+		return err
+	}
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := r.handlerFunc(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+
+	io.Copy(r.output, resp.Body)
+	io.WriteString(r.output, "\n")
+
+	return nil
+}
+
+// submit the set of metrics to datadog. submit will retry the operation if
+// unsuccessful
+func (r *Reporter) submit(metrics []*metric) {
+	s := series{Metrics: metrics}
+
+	data, err := json.Marshal(s)
+	if err != nil {
+		return
+	}
+
+	for attempts := 0; attempts < 3; attempts++ {
+		if err := r.post(data); err != nil {
+			fmt.Fprintln(os.Stderr, "failed to connect to host")
+			fmt.Fprintln(r.output, "failed to connect to host")
+			time.Sleep(time.Second * 15)
+			continue
+		}
+
+		break
+	}
+
+	for _, m := range metrics {
+		m.Reset()
+		poolMetric.Put(m)
+	}
+}
+
+// flush is a thread-unsafe flush. assumes caller has already obtained lock
+func (r *Reporter) flush() {
+	if r.offset == 0 {
+		return
+	}
+
+	var metrics []*metric
+	metrics = append(metrics, r.metrics[0:r.offset]...)
+	go r.submit(metrics)
+
+	for i := 0; i < r.offset; i++ {
+		r.metrics[i] = nil
+	}
+
+	r.offset = 0
+}
+
+// pushMetric provides a helper to push a metric onto the stack.  captures lock
+func (r *Reporter) pushMetric(m *metric) {
+	r.mux.Lock()
+	defer r.mux.Unlock()
+
+	r.metrics[r.offset] = m
+	r.offset++
+
+	if r.offset == r.bufSize {
```

Comment:
Do you need offset?  Can't you simply do a `len(r.metrics)`?  It's better to have one source of truth if possible.